### PR TITLE
add action envelope and sealed envelope

### PR DESCRIPTION
**Pull request by yutongp**
add action envelope to pack action
add sealed action envelope to authenticate an action

### CODE REVIEW

File: action/builder.go [hunk 1]
```diff
@@ -78,5 +79,74 @@ func (b *Builder) Build() AbstractAction {
...
+	b.elp.gasLimit = l
+	return b
+}
+
+// SetGasPrice sets action's gas price.
+func (b *EnvelopeBuilder) SetGasPrice(p *big.Int) *EnvelopeBuilder {
+	if p == nil {
+		return b
+	}
+	b.elp.gasPrice = &big.Int{}
```
Comment (dustinxie) [inline action/builder.go]: can simplify to = new(big.Int).Set(p)

File: action/builder.go [hunk 2]
```diff
@@ -78,5 +79,74 @@ func (b *Builder) Build() AbstractAction {
...
+	b.elp.gasPrice.Set(p)
+	return b
+}
+
+// SetGasPriceByBytes sets action's gas price from a byte slice source.
+func (b *EnvelopeBuilder) SetGasPriceByBytes(buf []byte) *EnvelopeBuilder {
+	if len(buf) == 0 {
+		return b
+	}
+	b.elp.gasPrice = &big.Int{}
```
Comment (dustinxie) [inline action/builder.go]: can simplify to = new(big.Int).SetBytes(buf)

File: action/execution_test.go [hunk 3]
```diff
@@ -26,73 +26,22 @@ func TestExecutionSignVerify(t *testing.T) {
...
+	elp := bd.SetNonce(0).
+		SetGasLimit(uint64(10)).
+		SetDestinationAddress(contract.RawAddress).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(ex).Build()
 
-	newex := &Execution{}
-	err = newex.Deserialize(s)
-	require.NoError(err)
+	w := AssembleSealedEnvelop(elp, executor.RawAddress, executor.PublicKey, []byte("lol"))
```
Comment (dustinxie) [inline action/execution_test.go]: nit: Envelope? missing last 'e'

File: action/startsubchain.go [hunk 4]
```diff
@@ -64,45 +63,22 @@ func NewStartSubChain(
...
-	start.operationDeposit = big.NewInt(0)
 	start.startHeight = startPb.StartHeight
 	start.parentHeightOffset = startPb.ParentHeightOffset
-	if len(startPb.SecurityDeposit) > 0 {
-		start.securityDeposit.SetBytes(startPb.SecurityDeposit)
-	}
-	if len(startPb.OperationDeposit) > 0 {
-		start.operationDeposit.SetBytes(startPb.OperationDeposit)
-	}
+	start.securityDeposit = big.NewInt(0)
```
Comment (dustinxie) [inline action/startsubchain.go]: this if() can be safely removed?

File: actpool/actpool.go [hunk 5]
```diff
@@ -139,12 +146,19 @@ func (ap *actPool) Add(act action.Action) error {
 	}
 	hash := act.Hash()
 	// Reject action if it already exists in pool
-	if ap.allActions[hash] != nil {
+	if _, exist := ap.allActions[hash]; exist {
```
Comment (dustinxie) [inline actpool/actpool.go]: good practice

File: blockchain/block.go [hunk 6]
```diff
@@ -69,7 +69,7 @@ type BlockFooter struct {
 // Block defines the struct of block
 type Block struct {
 	Header          *BlockHeader
-	Actions         []action.Action
+	Actions         []action.SealedEnvelope
 	SecretProposals []*action.SecretProposal
```
Comment (dustinxie) [inline blockchain/block.go]: add TODO about SecretProposal/Witness

File: blockchain/block.go [hunk 7]
```diff
@@ -260,72 +248,13 @@ func (b *Block) ConvertFromBlockHeaderPb(pbBlock *iproto.BlockPb) {
...
-			b.Actions = append(b.Actions, createDeposit)
-		} else if settleDepositPb := actPb.GetSettleDeposit(); settleDepositPb != nil {
-			settleDeposit := &action.SettleDeposit{}
-			if err := settleDeposit.LoadProto(actPb); err != nil {
-				return err
-			}
-			b.Actions = append(b.Actions, settleDeposit)
-		}
+		act := action.SealedEnvelope{}
+		act.LoadProto(actPb)
```
Comment (dustinxie) [inline blockchain/block.go]: zan!

File: action/action.go [hunk 8]
```diff
@@ -27,166 +27,310 @@ var (
...
+	stream = append(stream, payload...)
 	return stream
 }
 
-// Sign signs the action using sender's private key
-func Sign(act Action, sk keypair.PrivateKey) error {
-	// TODO: remove this conversion once we deprecate old address format
-	srcAddr, err := address.IotxAddressToAddress(act.SrcAddr())
+// Hash returns the hash value of SealedEnvelope.
+func (sealed SealedEnvelope) Hash() hash.Hash32B {
```
Comment (dustinxie) [inline action/action.go]: can this use *SealedEnvelope to be consistent?

File: action/action.go [hunk 9]
```diff
@@ -27,166 +27,310 @@ var (
...
+	stream := sealed.Envelope.ByteStream()
+	stream = append(stream, sealed.srcAddr...)
+	stream = append(stream, sealed.srcPubkey[:]...)
+	return blake2b.Sum256(stream)
+}
+
+// SrcAddr returns the source address
+func (sealed SealedEnvelope) SrcAddr() string { return sealed.srcAddr }
```
Comment (dustinxie) [inline action/action.go]: can this use *SealedEnvelope to be consistent?

File: action/action.go [hunk 10]
```diff
@@ -27,166 +27,310 @@ var (
...
+
+// SrcPubkey returns the source public key
+func (sealed SealedEnvelope) SrcPubkey() keypair.PublicKey { return sealed.srcPubkey }
```
Comment (dustinxie) [inline action/action.go]: can this use *SealedEnvelope to be consistent?

File: action/action.go [hunk 11]
```diff
@@ -27,166 +27,310 @@ var (
...
+	sealed := SealedEnvelope{Envelope: act}
+
 	// TODO: we should avoid generate public key from private key in each signature
 	pk, err := crypto.EC283.NewPubKey(sk)
 	if err != nil {
-		return errors.Wrapf(err, "error when deriving public key from private key")
+		return sealed, errors.Wrapf(err, "error when deriving public key from private key")
 	}
-	pkHash := keypair.HashPubKey(pk)
-	// TODO: abstract action shouldn't be aware that the playload is the hash of public key
```
Comment (dustinxie) [inline action/action.go]: is it possible to keep this in new code?
it is a safety/security check to make sure the action does come from signer

Comment (yutongp) [reply action/action.go]: no need to do that check any more. action doesn't tie to an identity til the signing.

File: action/startsubchain.go (hunk 4 above)
Comment (yutongp) [reply action/startsubchain.go]: yes

File: action/protocol/account/transfer.go [hunk 12]
```diff
@@ -80,7 +80,7 @@ func (p *Protocol) handleTransfer(act action.Action, raCtx protocol.RunActionsCt
 			return errors.Wrapf(err, "failed to update the Balance of sender %s", tsf.Sender())
 		}
 		// update sender Nonce
-		SetNonce(tsf, sender)
+		protocol.SetNonce(tsf, sender)
```
Comment (zjshen14) [inline action/protocol/account/transfer.go]: We probably don't want to move SetNonce to protocol package

File: action/protocol/execution/protocol_test.go [hunk 13]
```diff
@@ -347,10 +416,20 @@ func TestProtocol_Handle(t *testing.T) {
...
-		blk, err := bc.MintNewBlock([]action.Action{execution}, ta.Addrinfo["producer"],
+
+		bd := &action.EnvelopeBuilder{}
+		elp := bd.SetAction(execution).
+			SetDestinationAddress(action.EmptyAddress).
+			SetNonce(1).
+			SetGasLimit(10000000).Build()
+		selp, err := action.Sign(elp, ta.Addrinfo["producer"].RawAddress, ta.Addrinfo["producer"].PrivateKey)
+		require.NoError(err)
+		fmt.Println("XXXXXX", selp.GasLimit(), selp.Hash())
```
Comment (zjshen14) [inline action/protocol/execution/protocol_test.go]: Remove this line

File: action/protocol/execution/protocol_test.go [hunk 14]
```diff
@@ -347,10 +416,20 @@ func TestProtocol_Handle(t *testing.T) {
...
+
+		blk, err := bc.MintNewBlock([]action.SealedEnvelope{selp}, ta.Addrinfo["producer"],
 			nil, nil, "")
 		require.NoError(err)
+		fmt.Println("MMM", bc.ValidateBlock(blk, true))
```
Comment (zjshen14) [inline action/protocol/execution/protocol_test.go]: Remove this line

File: action/protocol/multichain/subchain/protocol.go [hunk 15]
```diff
@@ -11,9 +11,10 @@ import (
 	"fmt"
 	"math/big"
 
-	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
 	"github.com/pkg/errors"
 
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+
```
Comment (zjshen14) [inline action/protocol/multichain/subchain/protocol.go]: No empty line between

File: action/protocol/vote/protocol_test.go [hunk 16]
```diff
@@ -146,10 +146,12 @@ func TestProtocol_Validate(t *testing.T) {
 	protocol := NewProtocol(bc)
 
 	// Caes I: Oversized data
-	vote, err := action.NewVote(1, "src", "dst", uint64(100000), big.NewInt(0))
+	var dst string
+	for i := 0; i < 10000; i++ {
```
Comment (zjshen14) [inline action/protocol/vote/protocol_test.go]: what does this do?

File: action/action_test.go [hunk 17]
```diff
@@ -0,0 +1,38 @@
+package action
```
[the remark this answers is not in the document]
Comment (raullenchai) [reply action/action_test.go]: +1

File: action/protocol/vote/protocol_test.go (hunk 16 above)
Comment (yutongp) [reply action/protocol/vote/protocol_test.go]: create an oversize action

**Pull request merged**